File: backend/app/routes/ai.js
import express from 'express';
import User from '../models/User.js';
import AIUsage from '../models/AIUsage.js';
import { authMiddleware } from '../middleware/auth.js';

const router = express.Router();

const LIMITS = {
  starter: 20,
  growth: 200,
  premium: 1000,
  null: 5
};

router.post('/suggest', authMiddleware, async (req, res) => {
  try {
    const user = await User.findById(req.user.sub);
    if (!user) return res.status(404).json({ detail: 'User not found.' });

    const subTier = user.subscription;
    const limit = LIMITS[subTier] || 5;

    // Count calls this month
    const startOfMonth = new Date();
    startOfMonth.setDate(1);
    startOfMonth.setHours(0, 0, 0, 0);

    const count = await AIUsage.countDocuments({
      user_id: user._id,
      timestamp: { $gte: startOfMonth }
    });

    if (count >= limit) {
      return res.status(429).json({
        detail: `AI monthly limit reached (${count}/${limit} calls used). Please upgrade your subscription.`
      });
    }

    const { prompt, action_type } = req.body;
    let suggestion = 'Eximarg AI consultant recommended: Ensure all products match the correct ITC (HS) 8-digit code. In India, APEDA regulates agricultural food exports, while FIEO provides general RCMC certificates. Ensure your buyer conversation has a formal proforma invoice with clear FOB/CIF terms before shipping.';

    if (prompt?.toLowerCase().includes('reply') || action_type === 'improve_reply') {
      suggestion = 'Dear Buyer, Thank you for expressing interest in our premium export catalog. We have processed your inquiry and attached our detailed FOB pricing table and MOQ tiers. Let us know if you require custom sampling before the bulk deal. Best regards, EXIMARG Export Division.';
    } else if (prompt?.toLowerCase().includes('hsn')) {
      suggestion = 'We suggest reviewing the Spices Board or FIEO directories for the correct HSN code. Common Codes: 0904 (Pepper/Chilli), 1006 (Basmati Rice), 6205 (Cotton Woven Shirts).';
    }

    await AIUsage.create({
      user_id: user._id,
      action: action_type || 'consult',
      prompt
    });

    res.json({
      suggestion,
      calls_used: count + 1,
      limit
    });
  } catch (error) {
    res.status(500).json({ detail: error.message });
  }
});

export default router;

File: backend/app/routes/ai.py
from fastapi import APIRouter, HTTPException, Depends
from bson import ObjectId
from datetime import datetime
from app.services.db import get_db
from app.utils.helpers import get_current_user_payload

router = APIRouter(prefix="/api/ai", tags=["ai"])

LIMITS = {
    "starter": 20,
    "growth": 200,
    "premium": 1000,
    None: 5 # Default trial limit
}

@router.post("/suggest")
async def ai_suggest(payload: dict, current_user: dict = Depends(get_current_user_payload)):
    db = get_db()
    user_id = current_user["sub"]
    
    user = await db.users.find_one({"_id": ObjectId(user_id)})
    if not user:
        raise HTTPException(status_code=404, detail="User not found")
        
    sub_tier = user.get("subscription")
    limit = LIMITS.get(sub_tier, 5)
    
    # Calculate calls in current month
    now = datetime.utcnow()
    start_of_month = datetime(now.year, now.month, 1)
    
    count = await db.ai_usage.count_documents({
        "user_id": user_id,
        "timestamp": {"$gte": start_of_month}
    })
    
    if count >= limit:
        raise HTTPException(
            status_code=429,
            detail=f"AI monthly limit reached ({count}/{limit} calls used). Please upgrade your subscription."
        )
        
    prompt = payload.get("prompt", "")
    action_type = payload.get("action_type", "consult") # consult, improve_reply, suggest_docs
    
    # Mock AI response based on prompt topic
    suggestion = "Eximarg AI consultant recommended: Ensure all products match the correct ITC (HS) 8-digit code. In India, APEDA regulates agricultural food exports, while FIEO provides general RCMC certificates. Ensure your buyer conversation has a formal proforma invoice with clear FOB/CIF terms before shipping."
    if "reply" in prompt.lower() or action_type == "improve_reply":
        suggestion = "Dear Buyer, Thank you for expressing interest in our premium export catalog. We have processed your inquiry and attached our detailed FOB pricing table and MOQ tiers. Let us know if you require custom sampling before the bulk deal. Best regards, EXIMARG Export Division."
    elif "hsn" in prompt.lower():
        suggestion = "We suggest reviewing the Spices Board or FIEO directories for the correct HSN code. Common Codes: 0904 (Pepper/Chilli), 1006 (Basmati Rice), 6205 (Cotton Woven Shirts)."
        
    # Log usage
    await db.ai_usage.insert_one({
        "user_id": user_id,
        "action": action_type,
        "prompt": prompt,
        "timestamp": datetime.utcnow()
    })
    
    return {
        "suggestion": suggestion,
        "calls_used": count + 1,
        "limit": limit
    }
